File: barretenberg/test/wire_compat/cpp/echo_client.cpp
/**
 * Echo IPC client (C++) — uses GENERATED typed client.
 * Usage: echo_client --socket /tmp/echo.sock
 */

#include "generated/client_gen.hpp"
#include <iostream>
#include <cassert>
#include <string_view>

using namespace echo;

int main(int argc, char** argv) {
    const char* socket_path = nullptr;
    for (int i = 1; i < argc - 1; i++) {
        if (std::string_view(argv[i]) == "--socket") socket_path = argv[i + 1];
    }
    if (!socket_path) { std::cerr << "Usage: echo_client --socket <path>\n"; return 1; }

    EchoClient client(socket_path);

    // EchoBytes — using generated typed client method
    {
        EchoBytes cmd{ .data = { 0xDE, 0xAD, 0xBE, 0xEF, 0x42 } };
        auto resp = client.bytes(cmd);
        assert(resp.data == cmd.data);
        std::cerr << "echo_client(cpp): EchoBytes OK\n";
    }

    // EchoFields
    {
        EchoFields cmd{ .a = 42, .b = 999999, .name = "hello wire compat" };
        auto resp = client.fields(cmd);
        assert(resp.a == 42 && resp.b == 999999 && resp.name == "hello wire compat");
        std::cerr << "echo_client(cpp): EchoFields OK\n";
    }

    // EchoNested
    {
        EchoNested cmd; cmd.inner.values = { {1, 2, 3}, {4, 5} }; cmd.inner.flag = true;
        auto resp = client.nested(cmd);
        assert(resp.inner.values == cmd.inner.values && resp.inner.flag == cmd.inner.flag);
        std::cerr << "echo_client(cpp): EchoNested OK\n";
    }

    client.shutdown();
    std::cerr << "echo_client(cpp): all tests passed\n";
    return 0;
}

File: barretenberg/test/wire_compat/rust/src/echo_client.rs
//! Echo IPC client — connects, sends test commands, verifies responses.
//! Usage: echo_client --socket /tmp/echo.sock
//! Exits 0 on success, 1 on failure.

use echo_wire_compat::types_gen::*;
use echo_wire_compat::error::{EchoError, Result};
use std::io::{Read, Write};
use std::os::unix::net::UnixStream;

fn send_recv(stream: &mut UnixStream, command: Command) -> Result<Response> {
    // Serialize as [command] (tuple wrapper)
    let payload = rmp_serde::to_vec_named(&vec![command])
        .map_err(|e| EchoError::Serialization(e.to_string()))?;

    // Send length-prefixed
    let len = (payload.len() as u32).to_le_bytes();
    stream.write_all(&len)?;
    stream.write_all(&payload)?;
    stream.flush()?;

    // Read response length
    let mut len_buf = [0u8; 4];
    stream.read_exact(&mut len_buf)?;
    let resp_len = u32::from_le_bytes(len_buf) as usize;

    // Read response payload
    let mut resp_payload = vec![0u8; resp_len];
    stream.read_exact(&mut resp_payload)?;

    let response: Response = rmp_serde::from_slice(&resp_payload)
        .map_err(|e| EchoError::Deserialization(e.to_string()))?;

    Ok(response)
}

fn main() -> Result<()> {
    let args: Vec<String> = std::env::args().collect();
    let socket_path = args.iter()
        .position(|a| a == "--socket")
        .and_then(|i| args.get(i + 1))
        .expect("Usage: echo_client --socket <path>");

    let mut stream = UnixStream::connect(socket_path)?;

    // Test 1: EchoBytes
    let test_data = vec![0xDE, 0xAD, 0xBE, 0xEF, 0x42];
    let cmd = Command::EchoBytes(EchoBytes::new(test_data.clone()));
    match send_recv(&mut stream, cmd)? {
        Response::EchoBytesResponse(resp) => {
            assert_eq!(resp.data, test_data, "EchoBytes data mismatch");
            eprintln!("echo_client(rust): EchoBytes OK");
        }
        other => {
            eprintln!("echo_client(rust): EchoBytes unexpected response: {:?}", other);
            std::process::exit(1);
        }
    }

    // Test 2: EchoFields
    let cmd = Command::EchoFields(EchoFields::new(42, 999999, "hello wire compat".to_string()));
    match send_recv(&mut stream, cmd)? {
        Response::EchoFieldsResponse(resp) => {
            assert_eq!(resp.a, 42);
            assert_eq!(resp.b, 999999);
            assert_eq!(resp.name, "hello wire compat");
            eprintln!("echo_client(rust): EchoFields OK");
        }
        other => {
            eprintln!("echo_client(rust): EchoFields unexpected response: {:?}", other);
            std::process::exit(1);
        }
    }

    // Test 3: EchoNested
    let inner = EchoInner {
        values: vec![vec![1, 2, 3], vec![4, 5]],
        flag: Some(true),
    };
    let cmd = Command::EchoNested(EchoNested::new(inner.clone()));
    match send_recv(&mut stream, cmd)? {
        Response::EchoNestedResponse(resp) => {
            assert_eq!(resp.inner.values, inner.values);
            assert_eq!(resp.inner.flag, inner.flag);
            eprintln!("echo_client(rust): EchoNested OK");
        }
        other => {
            eprintln!("echo_client(rust): EchoNested unexpected response: {:?}", other);
            std::process::exit(1);
        }
    }

    // Shutdown
    let cmd = Command::EchoShutdown(EchoShutdown::new());
    let _ = send_recv(&mut stream, cmd);
    eprintln!("echo_client(rust): all tests passed");
    Ok(())
}
